src/marqo/core/document/tensor_fields_container.py

```diff
@@ -0,0 +1,342 @@
+
+        if not self.subfields:
+            return []
+
+        subfield_chunks = {subfield: self.subfields[subfield].sub_field_chunk for subfield in self.weights.keys()
+                           if subfield in self.subfields}
+        return [json.dumps(subfield_chunks)]
+
+    @property
+    def tensor_field_embeddings(self):
+        if self.embeddings:
+            # populated from existing tensor
+            return super().tensor_field_embeddings
+
+        if not self.subfields:
+            return []
+
+        combo_embeddings = [
+            np.array(self.subfields[subfield].sub_field_embedding) * weight for subfield, weight in self.weights.items()
+            if subfield in self.subfields
+        ]
+
+        vector_chunk = np.squeeze(np.mean(combo_embeddings, axis=0))
+        if self.normalize_embeddings:
+            # TODO check if the norm can be 0
+            vector_chunk = vector_chunk / np.linalg.norm(vector_chunk)
+
+        return [vector_chunk.tolist()]
+
+
+class TensorFieldsContainer:
+
+    def __init__(self, tensor_fields: List[str], custom_vector_fields: List[str], multimodal_combo_fields: dict):
+        self._tensor_field_map: Dict[str, Dict[str, TensorFieldContent]] = dict()
+        self._tensor_fields = set(tensor_fields)
+        self._custom_tensor_fields: Set[str] = set(custom_vector_fields)
+        self._multimodal_combo_fields = multimodal_combo_fields
+        self._multimodal_sub_field_reverse_map: Dict[str, Set[str]] = dict()
+
+        for field_name, weights in self._multimodal_combo_fields.items():
+            for sub_field in weights.keys():
+                if sub_field not in self._multimodal_sub_field_reverse_map:
+                    self._multimodal_sub_field_reverse_map[sub_field] = set()
+                self._multimodal_sub_field_reverse_map[sub_field].add(field_name)
+
+    def is_custom_tensor_field(self, field_name: str) -> bool:
+        return field_name in self._custom_tensor_fields
+
+    def is_multimodal_field(self, field_name: str) -> bool:
+        return field_name in self._multimodal_combo_fields
+
+    def get_multimodal_field_mapping(self, field_name: str) -> Optional[dict]:
+        return self._multimodal_combo_fields.get(field_name, None)
+
+    def get_multimodal_sub_fields(self) -> Set[str]:
+        return set(self._multimodal_sub_field_reverse_map.keys())
+
+    def remove_doc(self, doc_id: str):
+        if doc_id in self._tensor_field_map:
+            del self._tensor_field_map[doc_id]
+
+    def _add_tensor_field_content(self, doc_id: str, field_name: str, content: TensorFieldContent) -> None:
+        if doc_id not in self._tensor_field_map:
+            self._tensor_field_map[doc_id] = dict()
+        self._tensor_field_map[doc_id][field_name] = content
+
+    def tensor_fields_to_vectorise(self, *types: FieldType) -> Generator[str, str, TensorFieldContent]:
+        for doc_id in list(self._tensor_field_map.keys()):
```

Comment:
Passing these in the function call (pure function) makes your function more readable and reusable, though I admit it could make the code cluttered at times

Reply:
could you clarify a bit about what should be passed in in the function call?